File: src/main/java/com/rs/java/game/player/queue/PlayerActionQueue.java
package com.rs.java.game.player.queue;

import com.rs.core.tasks.WorldTask;
import com.rs.core.tasks.WorldTasksManager;
import com.rs.kotlin.game.player.queue.QueueType;

import java.util.ArrayDeque;
import java.util.Queue;

public final class PlayerActionQueue {

    private final Queue<QueuedAction> weakQueue = new ArrayDeque<>();
    private final Queue<QueuedAction> strongQueue = new ArrayDeque<>();

    private boolean running;

    public void enqueue(Runnable action) {
        weakQueue.add(new QueuedAction(action, 0));
        tryRunNext();
    }

    public void enqueue(int delay, Runnable action) {
        weakQueue.add(new QueuedAction(action, delay));
        tryRunNext();
    }

    public void enqueueStrong(Runnable action) {
        strongQueue.add(new QueuedAction(action, 0));
        tryRunNext();
    }

    public void enqueueDelay(int ticks) {
        weakQueue.add(new QueuedAction(null, ticks));
        tryRunNext();
    }

    private void tryRunNext() {
        if (running)
            return;

        QueuedAction next = !strongQueue.isEmpty()
                ? strongQueue.poll()
                : weakQueue.poll();

        if (next == null)
            return;

        running = true;

        if (next.delay > 0) {
            WorldTasksManager.schedule(new WorldTask() {
                @Override
                public void run() {
                    runAction(next);
                    stop();
                }
            }, next.delay);
        } else {
            runAction(next);
        }
    }

    private void runAction(QueuedAction action) {
        try {
            if (action.action != null)
                action.action.run();
        } catch (Throwable e) {
            e.printStackTrace();
        }

        running = false;
        tryRunNext();
    }

    public void clearWeak() {
        weakQueue.clear();
    }

    public void clearAll() {
        weakQueue.clear();
        strongQueue.clear();
    }

    private record QueuedAction(Runnable action, int delay) {}
}

File: src/raynna/game/player/queue/PlayerActionQueue.kt
package raynna.game.player.queue

import raynna.game.player.Player
import java.util.ArrayDeque

class PlayerActionQueue(
    private val player: Player,
) {
    private val queue = ArrayDeque<QueuedAction>()
    private var running: QueuedAction? = null

    fun enqueueWeak(action: Runnable) {
        queue.add(QueuedAction(action, 0, QueueType.WEAK))
    }

    fun enqueueNormal(action: Runnable) {
        queue.add(QueuedAction(action, 0, QueueType.NORMAL))
    }

    fun enqueueStrong(action: Runnable) {
        queue.add(QueuedAction(action, 0, QueueType.STRONG))
    }

    fun enqueueSoft(action: Runnable) {
        queue.add(QueuedAction(action, 0, QueueType.SOFT))
    }

    fun enqueueDelay(
        ticks: Int,
        type: QueueType = QueueType.NORMAL,
    ) {
        queue.add(QueuedAction(null, ticks, type))
    }

    fun process() {
        val current =
            running ?: run {
                if (queue.isEmpty()) return

                val next = queue.first()

                if (next.type == QueueType.STRONG) {
                    closeModalInterfaces()
                    removeWeakScripts()
                }

                if (!canRun(next)) return

                queue.removeFirst()
                running = next
                next
            }

        if (current.delay > 0) {
            current.delay--
            return
        }

        try {
            current.action?.run()
        } catch (e: Throwable) {
            e.printStackTrace()
        }

        running = null
    }

    private fun canRun(action: QueuedAction): Boolean {
        if (player.interfaceManager.containsChatBoxInter()) return false
        if (player.interfaceManager.containsScreenInter()) return false
        return true
    }

    private fun removeWeakScripts() {
        queue.removeIf { it.type == QueueType.WEAK }
    }

    private fun closeModalInterfaces() {
        player.interfaceManager.closeChatBoxInterface()
    }

    fun clearWeak() {
        queue.removeIf { it.type == QueueType.WEAK }
    }

    fun clearAll() {
        queue.clear()
        running = null
    }

    private data class QueuedAction(
        val action: Runnable?,
        var delay: Int,
        val type: QueueType,
    )
}
